File: contests/abc437/e.py
import sys
from collections import defaultdict

sys.setrecursionlimit(100000)

def main():
    n = int(input())
    target_arr = [tuple(map(int, input().split())) for _ in range(n)]
    children_dict_arr: list[dict[int, list[int]]] = [defaultdict(list) for _ in range(n+1)]
    for i, (x, y) in enumerate(target_arr):
        children_dict_arr[x][y].append(i+1)
    output = []
    listup([0], children_dict_arr, output)
    print(" ".join(map(str, output[1:])))
    
def listup(
    current_targets: list[int],
    children_dict_arr: list[dict[int, list[int]]],
    output: list[int]
):
    current_targets.sort()
    for current_target in current_targets:
        output.append(current_target)
    
    children_dict = defaultdict(list)
    for current_target in current_targets:
        unmarged_dict = children_dict_arr[current_target]
        for value, children_list in unmarged_dict.items():
            children_dict[value].extend(children_list)
    
    children_keys = list(children_dict.keys())
    children_keys.sort()
    for children_key in children_keys:
        listup(children_dict[children_key], children_dict_arr, output)
    


if __name__ == "__main__":
    main()

File: contests/abc437/e.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

void listup(
    std::vector<int64_t>& current_targets,
    std::vector<std::map<int64_t, std::vector<int64_t>>>& children_dict_arr,
    std::vector<int64_t>& output
) {
    std::sort(current_targets.begin(), current_targets.end());
    for (auto current_target: current_targets) { 
        output.push_back(current_target);
    }

    std::map<int64_t, std::vector<int64_t>> children_dict;
    for (auto current_target: current_targets) {
        for (auto [value, children_list]: children_dict_arr[current_target]) {
            for (auto children: children_list) {
                children_dict[value].push_back(children);
            }
        }
    }

    for (auto [key, value] : children_dict) {
        listup(value, children_dict_arr, output);
    }
    return;
}

int main() {
    int64_t n;
    std::cin >> n;
    std::vector<std::map<int64_t, std::vector<int64_t>>> children_dict_arr(n+1);
    for (int i=0; i<n; i++) {
        int64_t x, y;
        std::cin >> x >> y;
        children_dict_arr[x][y].push_back(i+1);
    }
    std::vector<int64_t> output;
    std::vector<int64_t> initial_targets = {0};
    listup(initial_targets, children_dict_arr, output);

    for (int i=1; i<n; i++) {
        std::cout << output[i] << " ";
    }
    std::cout << output[n] << std::endl;
    return 0;
}